### PR TITLE
[FLINK-23544][table-planner]Window TVF Supports session window

### PR DESCRIPTION
**Pull request by beyond1920**
## What is the purpose of the change

This pr is a subtask for support session window in Window TVF. This pr only focuses on planner, the next pr would focus on runtime.

## Brief change log

  - Introduce `SqlSessionTableFunction` and add a session table function object in `FlinkSqlOperatorTable`
  - Introduce `SessionWindowSpec`
  - Modify `WindowUtil#convertToWindowingStrategy` to add SESSION window
  - Disable `two-phase` (local-global) optimization for session window agg because window assigner results may be different
  - Disable `distinct-split`(partial-final) for session window agg because window assigner results may be different
  - Disable window table function for session window because `WindowTableFunctionOperator` is not suitable for Session Window because can't do state-less window assigning for input row per record for Session Window.

## Verifying this change

  - Add planner test for normal window aggregate on session window in `WindowAggregateTest`
  - Add planner test to validate two-phase optimization and distinct-phase optimization are disabled for window aggregate on session window in `WindowAggregateTest` 
  - Add planner test for validate window- assigner table function is disabled for session window

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlSessionTableFunction.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+public class SqlSessionTableFunction extends SqlWindowTableFunction {
+
+    public SqlSessionTableFunction() {
+        super(SqlKind.SESSION.name(), new OperandMetadataImpl());
+    }
+
+    /** Operand type checker for SESSION. */
+    private static class OperandMetadataImpl extends AbstractOperandMetadata {
+        OperandMetadataImpl() {
+            super(ImmutableList.of(PARAM_DATA, PARAM_TIMECOL, PARAM_STEP, PARAM_SIZE), 3);
```
Comment (tsreaper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlSessionTableFunction.java:46]: What does the 4th parameter do? It seems that a session window only needs 3 parameters.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/SplitAggregateRule.scala [hunk 2]
```diff
@@ -138,11 +139,14 @@ class SplitAggregateRule extends RelOptRule(
...
     val isProctimeWindowAgg = isWindowAgg && !windowProps.isRowtime
+    val isSessionWindowAgg = isWindowAgg &&
+      windowProps.getWindowSpec.isInstanceOf[SessionWindowSpec]
     // TableAggregate is not supported. see also FLINK-21923.
     val isTableAgg = AggregateUtil.isTableAggregate(agg.getAggCallList)
 
     agg.partialFinalType == PartialFinalType.NONE && agg.containsDistinctCall() &&
-      splitDistinctAggEnabled && isAllAggSplittable && !isProctimeWindowAgg && !isTableAgg
+      splitDistinctAggEnabled && isAllAggSplittable && !isProctimeWindowAgg &&
+      !isSessionWindowAgg && !isTableAgg
```
Comment (tsreaper) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/SplitAggregateRule.scala:149]: The logic seems to be quite messy here, and there is no explanation for the newly added session window.

Comment (beyond1920) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/SplitAggregateRule.scala]: add comment to explain for why disable distinct split for session window.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlSessionTableFunction.java (hunk 1 above)
Comment (beyond1920) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlSessionTableFunction.java]: good catch.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlSessionTableFunction.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            // check time attribute
+            return throwExceptionOrReturnFalse(
+                    checkTimeColumnDescriptorOperand(callBinding, 1), throwOnFailure);
+        }
+
+        @Override
+        public String getAllowedSignatures(SqlOperator op, String opName) {
+            return opName
+                    + "(TABLE table_name, DESCRIPTOR(timecol), "
+                    + "datetime interval, datetime interval)";
```
Comment (tsreaper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlSessionTableFunction.java:66]: Duplicated 4th parameter?

**Review by cshuo — COMMENTED** [review]
Thks beyond1920 for the efforts to make TVF window more complete. 

Routing runtime of session window to old window operator is a good idea,  IMO, there are two possible choices:
1. translate to old window operator in `StreamExecWindowAggregate` for session window (just as the PR);
2. directly translate to `StreamExecGroupWindowAggregate` in `StreamPhysicalWindowAggregate` for session window;

I think the 2nd choice may be better, as constructing an exec node is more concise and clean. 
Further, there are also other window cases that tvf-window can't handle, routing to old window exec node is more extensible.

**Comment by beyond1920** [comment]
@cshuo Thanks for advices, however I prefer the first choice fr the following reason:
1. It's better to have a clear separation from WindowAggregate and old GroupWindowAggregate in planner level. In the feature, we may deprecate GroupWindowAggregate directly. I don't want to introduce unnecessary coupling here.
2. Routine the session window TVF directly to `StreamExecGroupWindowAggregate ` would limit some cases for session window TVF. For example, Session window aggregate + WindowRank/ WindowJoin.
3. For cases which session window TVF currently not supported, I add strictly check in the planner, which would not lead to unexpected results.

What do you think?

**Comment by cshuo** [comment]
@beyond1920 My main concern is that `translateSessionWindowOperator` is actually duplicated and non-trivial code as that in `StreamExecGroupWindowAggregate`, I think we should avoid maintaining same code in two places.

For 2, I not quite clear about what you mean by "limit", routine to old window only changes the runtime operator, which does not affect any plan.

**Comment by beyond1920** [comment]
@cshuo, about point 2, if routine the session window TVF directly to StreamExecGroupWindowAggregate, some valid cases would not work. For example, Session window aggregate followed by WindowRank or WindowJoin. 
Because we only infer WindowProperties in `FlinkRelMdWindowProperties` for new Window TVF RelNodes which does not include legacy `GroupWindowAggregate`. BTW, I think there is no need to do this, either.

About the deduplicate code, they would not exist after we totally remove the GroupWindowAggregate in the later version.
I think it is OK, just like before we remove legacy planner, there are also deduplicate code in Blink planner and legacy planner. Besides, the code

**Comment by beyond1920** [comment]
@cshuo, Hi cshuo,  I'm misunderstand your meaning before, sorry about that. Your second choice is translate to `StreamExecGroupWindowAggregate` in `StreamPhysicalWindowAggregate` for session window. So it would not effect the planner compatible. 
I agree with you, the second choice would remove the code deduplication. After we deprecate the `GroupWindowAggregate `, we add current code back, they would be deduplicate code at that time.

**Comment by beyond1920** [comment]
@cshuo Hi cshuo, I've updated, please have a look, thanks.

**Comment by beyond1920** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamPhysicalWindowAggregate.scala [hunk 4]
```diff
@@ -104,14 +105,50 @@ class StreamPhysicalWindowAggregate(
...
-      namedWindowProperties.toArray,
-      InputProperty.DEFAULT,
-      FlinkTypeFactory.toLogicalRowType(getRowType),
-      getRelDetailedDescription
-    )
+    windowing.getWindow match {
+      case windowSpec: SessionWindowSpec =>
+        windowing match {
+          case timeWindowStrategy: TimeAttributeWindowingStrategy =>
+            val (timeAttributeFieldName, timeAttributeIdx) = if (timeWindowStrategy.isRowtime) {
```
Comment (cshuo) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamPhysicalWindowAggregate.scala:112]: branch here is not necessary, proctime case will be processed properly inside window op

**Review by cshuo — COMMENTED** [review]
LGTM, left minor comments.

**Comment by cshuo** [comment]
+1

**Comment by beyond1920** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
+1, will merge

**Pull request closed without merging**